oxidation/libparsec/crates/testbed/src/types.rs

```diff
@@ -0,0 +1,206 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
+
+use super::*;
+
+#[derive(Debug, Clone, serde::Serialize)]
+#[non_exhaustive] // Force use of `new()` constructor to compute crc
+pub struct TestbedTemplate {
+    pub id: &'static str,
+    pub root_signing_key: SigningKey,
+    pub devices: Vec<TestbedDeviceData>,
+    pub users: Vec<TestbedUserData>,
+    pub device_files: Vec<TestbedDeviceFileData>,
+    // TODO: finish me !
+    // pub manifests: Vec<std::sync::Arc<Manifest>>,
+    // pub vlobs: Vec<(VlobID, Vec<u8>)>,
+    // pub blocks: Vec<(BlockID, Vec<u8>)>,
+    // pub messages: Vec<(DeviceID, DateTime, Vec<u8>, MessageContent)>,
+    // pub sequester_services: Vec<(SequesterServiceID, SequesterServiceCertificate)>,
+    pub crc: u32,
+}
+
+impl TestbedTemplate {
+    pub fn device(&self, device_id: &DeviceID) -> &TestbedDeviceData {
+        self.devices
+            .iter()
+            .find(|x| x.device_id == *device_id)
+            .unwrap()
+    }
+    pub fn user(&self, user_id: &UserID) -> &TestbedUserData {
+        self.users.iter().find(|x| x.user_id == *user_id).unwrap()
+    }
+
+    pub fn new(
+        id: &'static str,
+        root_signing_key: SigningKey,
+        devices: Vec<TestbedDeviceData>,
+        users: Vec<TestbedUserData>,
+        device_files: Vec<TestbedDeviceFileData>,
+    ) -> Self {
+        let mut hasher = crc32fast::Hasher::new();
+
+        // Remember that changing the order (or adding items) to the hasher change the result !
```

Comment:
That seems normal behavior for an hash function (like `sha256` & co), I'm missing something ?
Maybe the reason of this comment is if this code block is change we may have to ensure the client & server run the same compiled code :thinking:

Reply:
> Maybe the reason of this comment is if this code block is change we may have to ensure the client & server run the same compiled code 🤔

Yes, basically changing the order is a breaking change that is not visible as long as the same codebase is used for the testbed server and the tests. As you say changing order for hash is pretty common, but we want to emphases anyway given how simple it is to add a component or reorder something here.